src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -210,11 +225,20 @@ public bool TryCancel()
                     switch ((State)state)
                     {
                         case State.Running:
-                            // A completion attempt is in progress. Keep busy-waiting.
-                            Trace("Busy wait");
-                            spinWait.SpinOnce();
+                            if (!abortIfRunning)
+                            {
+                                // A completion attempt is in progress. Keep busy-waiting.
+                                Trace("Busy wait");
+                                spinWait.SpinOnce();
+                            }
+                            else if (Interlocked.CompareExchange(ref _state, (int)State.Aborted, (int)State.Running) == (int)State.Running)
+                            {
+                                // We have successfully transitioned from Running to Aborted, break out of the loop.
+                                // Otherwise, the next iteration of the loop will handle the new state.
+                                Trace("Operation aborted");
+                                keepWaiting = false;
+                            }
```

Comment:
I tried splitting this switch statement up in the two things it is doing: change state, and exit the function if the oldstate was already completed. wdyt?

```cs
State oldState;
if (aborting)
{
    oldState = (State)Interlocked.Exchange(ref _state, (int)State.Aborted);
}
else
{
    // Cancel operation when it is not completed.
    SpinWait spinWait = default;
    do
    {
        oldState = (State)Interlocked.CompareExchange(ref _state, (int)State.Cancelled, (int)State.Waiting);

        if (oldState == State.Running)
        {
            // A completion attempt is in progress. Keep busy-waiting.
            Trace("Busy wait");
            spinWait.SpinOnce();
        }
    } while (oldState == State.Running);
}

// Exit if the operation had already completed.
switch (oldState)
{
    case State.Complete:
        Trace("Exit, previously completed");
        return false;
    case State.Cancelled:
    case State.Aborted:
        Trace("Exit, previously cancelled/aborted");
        return false;
}
```

Reply:
@tmds this looks better thanks, i'll make the changes asap.

there's just one thing i'm not sure about, which still revolves around the definition of `Aborted` vs `Cancelled` as mentioned [here](https://github.com/dotnet/runtime/pull/381/files/dd0ccf0469ae53707a0fb4e456dac0abc8e3b08f#r354797209):

When [aborting](https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1111) there'll be at most one operation on the queue that's actually in the `Running` state. The others will all be in `Waiting` state. So, with the current code, only the running operation will be aborted and the others will be cancelled which is consistent with the above [definition](https://github.com/dotnet/runtime/pull/381/files/dd0ccf0469ae53707a0fb4e456dac0abc8e3b08f#r354797209) : The running operation was aborted since it was "ungracefully" stopped while the waiting operations on the queue would be cancelled since they've not started to execute.